src/mlpack/core/data/load_numeric.hpp

```diff
@@ -13,117 +13,156 @@
 #ifndef MLPACK_CORE_DATA_LOAD_NUMERIC_HPP
 #define MLPACK_CORE_DATA_LOAD_NUMERIC_HPP
 
+#include "text_options.hpp"
+#include "load_categorical.hpp"
+#include "load_dense.hpp"
+#include "load_sparse.hpp"
+
 namespace mlpack {
 namespace data {
 
-// These help assemble the Armadillo csv_opts flags.
-inline const arma::csv_opts::opts& NoTransposeOpt(const bool noTranspose)
-{
-  if (noTranspose)
-    return arma::csv_opts::none;
-  else
-    return arma::csv_opts::trans;
-}
-
-inline const arma::csv_opts::opts& HasHeadersOpt(const bool hasHeaders)
-{
-  if (hasHeaders)
-    return arma::csv_opts::with_header;
-  else
-    return arma::csv_opts::no_header;
-}
-
-inline const arma::csv_opts::opts& SemicolonOpt(const bool semicolon)
+template<typename MatType>
+bool LoadNumeric(const std::string& filename,
+                 MatType& matrix,
+                 std::fstream& stream,
+                 TextOptions& opts)
 {
-  if (semicolon)
-    return arma::csv_opts::semicolon;
-  else
-    return arma::csv_opts::none;
-}
+  bool success = false;
 
-inline const arma::csv_opts::opts& MissingToNanOpt(const bool missingToNan)
-{
-  if (missingToNan)
+  TextOptions txtOpts(std::move(opts));
+  if constexpr (IsSparseMat<MatType>::value)
+  {
+    success = LoadSparse(filename, matrix, txtOpts, stream);
+  }
+  else if (txtOpts.Categorical() ||
+      (txtOpts.Format() == FileType::ARFFASCII))
+  {
+    success = LoadCategorical(filename, matrix, txtOpts);
+  }
+  else if constexpr (IsCol<MatType>::value)
   {
-    #if ARMA_VERSION_MAJOR >= 12
-    return arma::csv_opts::strict;
-    #else
-    return arma::csv_opts::none;
-    #endif
+    success = LoadCol(filename, matrix, txtOpts, stream);
+  }
+  else if constexpr (IsRow<MatType>::value)
+  {
+    success = LoadRow(filename, matrix, txtOpts, stream);
+  }
+  else if constexpr (IsDense<MatType>::value)
+  {
+    success = LoadDense(filename, matrix, txtOpts, stream);
   }
   else
   {
-    return arma::csv_opts::none;
+    return HandleError("data::Load(): unknown matrix-like type given!",
+        txtOpts);
   }
+  opts = std::move(txtOpts);
+  return success;
 }
 
 template<typename MatType>
-bool LoadCSVASCII(const std::string& filename,
-                  MatType& matrix,
-                  TextOptions& opts)
+bool LoadNumericVector(const std::vector<std::string>& filenames,
```

Comment:
This name could be a little confusing because you mean `Vector` in the sense of, we are loading many files, but it could also seem like the same thing as `LoadCol()` or `LoadRow()`.  I know it is minor because this is an internal function but I might consider a name more like `LoadNumericMultifile()` or, well, I don't know if that's actually an improvement.